crates/napi/src/threadsafe_function.rs

```diff
@@ -298,20 +302,24 @@ impl<T: 'static> ThreadsafeFunction<T, ErrorStrategy::Fatal> {
     if self.aborted.load(Ordering::Acquire) {
       return Status::Closing;
     }
-    unsafe {
+    let status = unsafe {
       sys::napi_call_threadsafe_function(
         self.raw_tsfn,
         Box::into_raw(Box::new(value)) as *mut _,
         mode.into(),
       )
     }
-    .into()
+    .into();
+    if status == Status::Closing {
```

Comment:
Unfortinately, I think, `AtomicBool` won't cut it here, you'll need a proper lock. `ThreadsafeFunction::call` can be called concurrently, and it is possible that the second call will happen after you called `napi_call_threadsafe_function`, but before you checked the result. In case you received `napi_closing` as the result of the first call, doing the second call is already unsafe.
My original reproduction repo used `RwLock` here, but we later switched to `Mutex` for exactly this reason. If you need it, you can find updated version [here](https://github.com/prisma/prisma-engines/blob/9e513c70f3a00c2f3ebf3115f2545ff2de8bce20/query-engine/query-engine-node-api/src/log_callback.rs).

Reply:
I use `Atomic` here rather than `Mutex` is because there is already `Mutex` in the `napi_call_threadsafe_function` in the Node.js; I don't want to introduce redundant overhead here.

Let me extrapolate the memory ordering here. I'll be back here when I have a conclusion 👀